libsolidity/ast/Types.cpp

```diff
@@ -2140,12 +2140,21 @@ FunctionType const* ContractType::newExpressionType() const
 	return m_constructorType;
 }
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables() const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(DataLocation _location) const
 {
+	VariableDeclaration::Location location = [&]() {
+		switch (_location)
+		{
+			case DataLocation::Storage: return VariableDeclaration::Location::Unspecified; // By default state variables have unspecified (storage) data location
+			case DataLocation::Transient: return VariableDeclaration::Location::Transient;
+			default: solAssert(false);
+		}
+	}();
+
 	std::vector<VariableDeclaration const*> variables;
 	for (ContractDefinition const* contract: m_contract.annotation().linearizedBaseContracts | ranges::views::reverse)
 		for (VariableDeclaration const* variable: contract->stateVariables())
-			if (!(variable->isConstant() || variable->immutable()))
+			if (!(variable->isConstant() || variable->immutable()) && variable->referenceLocation() == location)
```

Comment:
Constants and immutables can also be unspecified, and they are not stored anywhere.
If we refactor as this suggested [here](https://github.com/ethereum/solidity/pull/15247/files#r1671291082), then I think we can remove the check.

Reply:
Constants and immutables are excluded in the first conditional `!(variable->isConstant() || variable->immutable())`, so that's not an issue. The question on my side was bad though - the reason you have this check is to filter storage vs transient storage state variables when generating layouts, so this check is necessary regardless.